File: src/injector.ts
// Port of internal/injector/injector.go — pure string manipulation, no Obsidian API.

export type AttachmentMode = 'vault-root' | 'fixed-folder' | 'same-as-note' | 'subfolder';

export interface AttachmentConfig {
  mode: AttachmentMode;
  folderName: string; // used by fixed-folder and subfolder modes
}

/**
 * Parses Obsidian's attachmentFolderPath setting into a structured config.
 * Mirrors parseAttachmentFolderPath in main.go.
 */
export function parseAttachmentFolderPath(p: string | undefined): AttachmentConfig {
  if (!p) return { mode: 'vault-root', folderName: '' };

  if (p.startsWith('/')) {
    return { mode: 'fixed-folder', folderName: p.slice(1) };
  }
  if (p === './' || p === '.') {
    return { mode: 'same-as-note', folderName: '' };
  }
  if (p.startsWith('./')) {
    return { mode: 'subfolder', folderName: p.slice(2) };
  }
  // Plain name without prefix — treat as fixed folder.
  return { mode: 'fixed-folder', folderName: p };
}

/**
 * Returns the vault-relative destination directory for a given note.
 * noteDir is the vault-relative directory of the note (empty string = vault root).
 */
export function resolveDestDir(cfg: AttachmentConfig, noteDir: string, vaultRoot = ''): string {
  switch (cfg.mode) {
    case 'vault-root':    return vaultRoot;
    case 'fixed-folder':  return cfg.folderName;
    case 'same-as-note':  return noteDir;
    case 'subfolder':     return noteDir ? `${noteDir}/${cfg.folderName}` : cfg.folderName;
  }
}

// ── Annotation injection ──────────────────────────────────────────────────────

/**
 * Builds the annotation string for a successfully downloaded file.
 * Format: ([[excalidraw-ID.excalidraw|local copy DD/MM/YYYY]])
 */
export function buildAnnotation(filename: string): string {
  const now = new Date();
  const dd = String(now.getDate()).padStart(2, '0');
  const mm = String(now.getMonth() + 1).padStart(2, '0');
  const yyyy = now.getFullYear();
  return `([[${filename}|local copy ${dd}/${mm}/${yyyy}]])`;
}

export const ANNOTATION_WARNING       = '(⚠ download failed)';
export const ANNOTATION_REFRESH_FAILED = '(⚠ refresh failed)';
export const ANNOTATION_EMPTY_CANVAS  = '(⚠ empty canvas — ask the author to open the room, then re-run the downloader)';

/**
 * Checks if the first occurrence of url in content already has a local-copy annotation.
 * Used in refresh mode to decide whether to preserve the existing annotation on failure.
 */
export function hasExistingLocalCopy(content: string, url: string): boolean {
  const idx = content.indexOf(url);
  if (idx === -1) return false;
  let after = content.slice(idx + url.length);
  if (after.startsWith(')')) after = after.slice(1);
  return after.replace(/^[ \t]+/, '').startsWith('([[');
}

/**
 * Applies all url→annotation replacements to content.
 * URLs absent from the map are left untouched.
 */
export function injectAll(
  content: string,
  annotations: Map<string, string>,
  refresh: boolean,
): string {
  for (const [url, annotation] of annotations) {
    content = injectOne(content, url, annotation, refresh);
  }
  return content;
}

/**
 * Inserts annotation after every occurrence of url in content.
 * Handles bare URLs and markdown [text](url) links.
 * Is idempotent: skips occurrences already followed by ([[.
 * In refresh mode, replaces existing ([[...]])  annotations.
 *
 * Mirrors injectOne in injector.go exactly.
 */
export function injectOne(
  content: string,
  url: string,
  annotation: string,
  refresh: boolean,
): string {
  let result = '';
  let remaining = content;

  for (;;) {
    const idx = remaining.indexOf(url);
    if (idx === -1) {
      result += remaining;
      break;
    }

    result += remaining.slice(0, idx + url.length);
    let after = remaining.slice(idx + url.length);

    // Consume closing ) if inside a markdown link [text](url)
    let closeParen = '';
    if (after.startsWith(')')) {
      closeParen = ')';
      after = after.slice(1);
    }

    const trimmed = after.replace(/^[ \t]+/, '');

    if (trimmed.startsWith('([[')) {
      if (refresh && annotation === ANNOTATION_REFRESH_FAILED) {
        // Refresh failed but old local copy exists: keep ([[...]]) and append error marker.
        const end = trimmed.indexOf(']])');
        if (end >= 0) {
          const existing = trimmed.slice(0, end + 3); // ([[...|local copy DATE]])
          after = trimmed.slice(end + 3);
          const rest = after.replace(/^[ \t]+/, '');
          if (rest.startsWith(ANNOTATION_REFRESH_FAILED)) {
            // Already marked — leave as-is
            result += closeParen + ' ' + existing;
            after = rest.slice(ANNOTATION_REFRESH_FAILED.length);
          } else {
            result += closeParen + ' ' + existing + ' ' + ANNOTATION_REFRESH_FAILED;
          }
        } else {
          result += closeParen;
        }
      } else if (refresh) {
        // Successful refresh: replace ([[...]]) and remove any trailing (⚠ refresh failed).
        const end = trimmed.indexOf(']])');
        if (end >= 0) {
          after = trimmed.slice(end + 3);
          const rest = after.replace(/^[ \t]+/, '');
          if (rest.startsWith(ANNOTATION_REFRESH_FAILED)) {
            after = rest.slice(ANNOTATION_REFRESH_FAILED.length);
          }
        }
        result += closeParen + ' ' + annotation;
      } else {
        // Normal mode: already annotated — leave as-is
        result += closeParen;
      }
    } else if (trimmed.startsWith('([local copy')) {
      // Old markdown format — migrate to wikilink
      const inner = trimmed.slice(1); // strip leading (
      const end = indexAfterMarkdownLink(inner);
      if (end >= 0 && end < inner.length && inner[end] === ')') {
        after = inner.slice(end + 1);
      }
      result += closeParen + ' ' + annotation;
    } else if (trimmed.startsWith('[local copy')) {
      // Old bare format — migrate to wikilink
      const end = indexAfterMarkdownLink(trimmed);
      if (end >= 0) after = trimmed.slice(end);
      result += closeParen + ' ' + annotation;
    } else if (trimmed.startsWith('(⚠')) {
      // Previous attempt failed — replace warning with new annotation
      const end = trimmed.indexOf(')');
      if (end >= 0) after = trimmed.slice(end + 1);
      result += closeParen + ' ' + annotation;
    } else {
      result += closeParen + ' ' + annotation;
    }

    remaining = after;
  }

  return result;
}

/**
 * Returns the index just past the closing ')' of the first markdown link
 * "[text](url)" in s, or -1 if not found.
 * Mirrors indexAfterMarkdownLink in injector.go.
 */
function indexAfterMarkdownLink(s: string): number {
  const closeBracket = s.indexOf('](');
  if (closeBracket === -1) return -1;
  const rest = s.slice(closeBracket + 2);
  const closeParenIdx = rest.indexOf(')');
  if (closeParenIdx === -1) return -1;
  return closeBracket + 2 + closeParenIdx + 1;
}

File: cli/internal/injector/injector.go
package injector

import (
	"errors"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"

	"obsidian-excalidraw-downloader/internal/excalidraw"
)

const (
	annotationWarning        = "(⚠ download failed)"
	annotationEmptyCanvas    = "(⚠ empty canvas — ask the author to open the room, then re-run the downloader)"
	annotationRefreshFailed  = "(⚠ refresh failed)"
)

// AttachmentMode mirrors Obsidian's "Default location for new attachments" setting.
type AttachmentMode int

const (
	// ModeVaultRoot saves attachments to the vault root (attachmentFolderPath = "").
	ModeVaultRoot AttachmentMode = iota
	// ModeFixedFolder saves attachments to a fixed folder in the vault (attachmentFolderPath = "/folder").
	ModeFixedFolder
	// ModeSameAsNote saves attachments next to each note (attachmentFolderPath = "./").
	ModeSameAsNote
	// ModeSubfolder saves attachments in a subfolder relative to each note (attachmentFolderPath = "./folder").
	ModeSubfolder
)

// Processor handles finding and annotating Excalidraw links in markdown files.
type Processor struct {
	vaultPath  string
	mode       AttachmentMode
	folderName string // used by ModeFixedFolder and ModeSubfolder
	refresh    bool   // when true, re-downloads existing files and updates annotations
	log        *log.Logger
}

func New(vaultPath string, mode AttachmentMode, folderName string, refresh bool, logger *log.Logger) *Processor {
	return &Processor{
		vaultPath:  vaultPath,
		mode:       mode,
		folderName: folderName,
		refresh:    refresh,
		log:        logger,
	}
}

// resourcesDirFor returns the attachment directory for a given note path.
func (p *Processor) resourcesDirFor(notePath string) string {
	switch p.mode {
	case ModeVaultRoot:
		return p.vaultPath
	case ModeFixedFolder:
		return filepath.Join(p.vaultPath, p.folderName)
	case ModeSameAsNote:
		return filepath.Dir(notePath)
	case ModeSubfolder:
		return filepath.Join(filepath.Dir(notePath), p.folderName)
	}
	return p.vaultPath
}

// ProcessFile finds Excalidraw links in a markdown file, downloads them, and
// injects annotations (local copy link or warning) next to each URL.
// Returns counts for: newly downloaded, already cached on disk, empty canvas,
// errors, and (in refresh mode) links that failed to refresh but kept their
// existing local copy unchanged.
func (p *Processor) ProcessFile(path string) (downloaded, cached, empty, errCount, refreshFailed int) {
	data, err := os.ReadFile(path)
	if err != nil {
		p.log.Printf("  ERROR reading %s: %v", relPath(p.vaultPath, path), err)
		return 0, 0, 0, 1, 0
	}

	content := string(data)
	links := excalidraw.ParseAll(content, p.refresh)
	if len(links) == 0 {
		return 0, 0, 0, 0, 0
	}

	p.log.Printf("FILE %s — %d link(s) found", relPath(p.vaultPath, path), len(links))

	resourcesDir := p.resourcesDirFor(path)
	if err := os.MkdirAll(resourcesDir, 0755); err != nil {
		p.log.Printf("  ERROR creating resources dir %s: %v", resourcesDir, err)
		return 0, 0, 0, len(links), 0
	}

	// In refresh mode, track which URLs already had a local copy annotation so
	// that a failed re-download leaves those annotations untouched.
	hadLocalCopy := make(map[string]bool, len(links))
	if p.refresh {
		for _, link := range links {
			hadLocalCopy[link.URL] = hasExistingLocalCopy(content, link.URL)
		}
	}

	annotations := make(map[string]string, len(links))
	for _, link := range links {
		localFile, wasCached, dlErr := p.downloadTo(link, resourcesDir)
		if dlErr != nil {
			if p.refresh && hadLocalCopy[link.URL] {
				// New download failed but the old file is still intact —
				// keep existing annotation and append refresh failure marker.
				p.log.Printf("  REFRESH FAIL [%s] %s: %v — keeping existing copy", link.Kind, link.ID, dlErr)
				annotations[link.URL] = annotationRefreshFailed
				refreshFailed++
			} else if errors.Is(dlErr, excalidraw.ErrEmptyCanvas) {
				p.log.Printf("  EMPTY [room] %s: canvas empty — data may exist on another browser", link.ID)
				annotations[link.URL] = annotationEmptyCanvas
				empty++
			} else {
				p.log.Printf("  ERROR [%s] %s: %v", link.Kind, link.ID, dlErr)
				annotations[link.URL] = annotationWarning
				errCount++
			}
		} else {
			filename := filepath.Base(localFile)
			date := time.Now().Format("02/01/2006")
			annotation := fmt.Sprintf("([[%s|local copy %s]])", filename, date)
			annotations[link.URL] = annotation
			if wasCached {
				p.log.Printf("  CACHE [%s] %s → %s", link.Kind, link.ID, filename)
				cached++
			} else {
				p.log.Printf("  OK    [%s] %s → %s", link.Kind, link.ID, filename)
				downloaded++
			}
		}
	}

	newContent := injectAll(content, annotations, p.refresh)
	if newContent == content {
		return downloaded, cached, empty, errCount, refreshFailed
	}

	if err := os.WriteFile(path, []byte(newContent), 0644); err != nil {
		p.log.Printf("  ERROR writing %s: %v", relPath(p.vaultPath, path), err)
		return downloaded, cached, empty, errCount + 1, refreshFailed
	}

	return downloaded, cached, empty, errCount, refreshFailed
}

// hasExistingLocalCopy reports whether the first occurrence of url in content
// is already followed by a ([local copy annotation.
func hasExistingLocalCopy(content, url string) bool {
	idx := strings.Index(content, url)
	if idx == -1 {
		return false
	}
	after := content[idx+len(url):]
	if len(after) > 0 && after[0] == ')' {
		after = after[1:]
	}
	return strings.HasPrefix(strings.TrimLeft(after, " \t"), "([[")
}

// downloadTo calls excalidraw.Download with retry logic:
//   - Room links: retry indefinitely on transient errors (timeout, connection).
//     ErrEmptyCanvas is not retried — it means the data is on another machine.
//   - Non-room links: single attempt, failures are permanent.
//
// Returns (localPath, cached, err) where cached is true when the file already
// existed on disk and no network request was made.
func (p *Processor) downloadTo(link excalidraw.Link, destDir string) (string, bool, error) {
	if link.Kind != excalidraw.Room {
		return excalidraw.Download(link, destDir, p.refresh)
	}

	for attempt := 1; ; attempt++ {
		localFile, cached, err := excalidraw.Download(link, destDir, p.refresh)
		if err == nil {
			return localFile, cached, nil
		}
		if errors.Is(err, excalidraw.ErrEmptyCanvas) {
			return "", false, err // not retryable: data is on another participant's browser
		}
		p.log.Printf("  RETRY [room] %s (attempt %d): %v — retrying in 5s...", link.ID, attempt, err)
		time.Sleep(5 * time.Second)
	}
}

// injectAll applies all URL→annotation pairs to content.
func injectAll(content string, annotations map[string]string, refresh bool) string {
	for url, annotation := range annotations {
		content = injectOne(content, url, annotation, refresh)
	}
	return content
}

// injectOne inserts annotation after each occurrence of url in content.
// Handles both bare URLs and markdown links [text](url).
// Is idempotent: skips occurrences that already have a ([local copy annotation,
// unless refresh is true — in that case the existing annotation is replaced.
func injectOne(content, url, annotation string, refresh bool) string {
	var b strings.Builder
	remaining := content

	for {
		idx := strings.Index(remaining, url)
		if idx == -1 {
			b.WriteString(remaining)
			break
		}

		b.WriteString(remaining[:idx+len(url)])
		after := remaining[idx+len(url):]

		// If the URL is inside a markdown link [text](URL), consume the closing ).
		closeParen := ""
		if len(after) > 0 && after[0] == ')' {
			closeParen = ")"
			after = after[1:]
		}

		// Check if already annotated right after the link.
		trimmed := strings.TrimLeft(after, " \t")
		if strings.HasPrefix(trimmed, "([[") {
			if refresh && annotation == annotationRefreshFailed {
				// Refresh failed but old local copy exists: keep ([[...]]) and append error marker.
				if end := strings.Index(trimmed, "]])"); end >= 0 {
					existing := trimmed[:end+3]
					after = trimmed[end+3:]
					rest := strings.TrimLeft(after, " \t")
					if strings.HasPrefix(rest, annotationRefreshFailed) {
						// Already marked — leave as-is.
						b.WriteString(closeParen)
						b.WriteString(" ")
						b.WriteString(existing)
					} else {
						b.WriteString(closeParen)
						b.WriteString(" ")
						b.WriteString(existing)
						b.WriteString(" ")
						b.WriteString(annotationRefreshFailed)
					}
				} else {
					b.WriteString(closeParen)
				}
			} else if refresh {
				// Replace existing wikilink annotation ([[file|local copy DATE]]).
				// Strip any trailing (⚠ refresh failed) if present.
				if end := strings.Index(trimmed, "]])"); end >= 0 {
					after = trimmed[end+3:]
				}
				rest := strings.TrimLeft(after, " \t")
				if strings.HasPrefix(rest, annotationRefreshFailed) {
					after = rest[len(annotationRefreshFailed):]
				}
				b.WriteString(closeParen)
				b.WriteString(" ")
				b.WriteString(annotation)
			} else {
				// Already annotated — leave as-is.
				b.WriteString(closeParen)
			}
		} else if strings.HasPrefix(trimmed, "([local copy") {
			// Old markdown format — migrate to wikilink.
			inner := trimmed[1:] // strip leading (
			if end := indexAfterMarkdownLink(inner); end >= 0 && end < len(inner) && inner[end] == ')' {
				after = inner[end+1:]
			}
			b.WriteString(closeParen)
			b.WriteString(" ")
			b.WriteString(annotation)
		} else if strings.HasPrefix(trimmed, "[local copy") {
			// Old bare format — migrate to wikilink.
			if endIdx := indexAfterMarkdownLink(trimmed); endIdx >= 0 {
				after = trimmed[endIdx:]
			}
			b.WriteString(closeParen)
			b.WriteString(" ")
			b.WriteString(annotation)
		} else if strings.HasPrefix(trimmed, "(⚠") {
			// Previous attempt failed — replace old warning with new annotation.
			if endIdx := strings.Index(trimmed, ")"); endIdx >= 0 {
				after = trimmed[endIdx+1:]
			}
			b.WriteString(closeParen)
			b.WriteString(" ")
			b.WriteString(annotation)
		} else {
			b.WriteString(closeParen)
			b.WriteString(" ")
			b.WriteString(annotation)
		}

		remaining = after
	}

	return b.String()
}

// indexAfterMarkdownLink returns the index just past the closing ')' of the
// first markdown link "[text](url)" in s, or -1 if not found.
func indexAfterMarkdownLink(s string) int {
	closeBracket := strings.Index(s, "](")
	if closeBracket == -1 {
		return -1
	}
	rest := s[closeBracket+2:]
	closeParenIdx := strings.Index(rest, ")")
	if closeParenIdx == -1 {
		return -1
	}
	return closeBracket + 2 + closeParenIdx + 1
}

func relPath(base, path string) string {
	r, err := filepath.Rel(base, path)
	if err != nil {
		return path
	}
	return r
}
